src/Microsoft.VisualStudio.Composition/PartDiscovery.cs

```diff
@@ -202,13 +202,46 @@ protected internal static Type GetTypeIdentityFromImportingType(Type type, bool
             return type;
         }
 
+        protected internal static TypeRef GetTypeIdentityFromImportingTypeRef(TypeRef typeRef, bool importMany)
+        {
+            Requires.NotNull(typeRef, nameof(typeRef));
+
+            if (importMany)
+            {
+                typeRef = typeRef.ElementTypeRef;
+            }
+
+            if (typeRef.IsAnyLazyType() || typeRef.IsExportFactoryTypeV1() || typeRef.IsExportFactoryTypeV2())
+            {
+                return typeRef.GenericTypeArguments[0];
+            }
+
+            return typeRef;
+        }
+
         protected internal static Type GetElementTypeFromMany(Type type)
         {
             Requires.NotNull(type, nameof(type));
 
+            if (TryGetElementTypeFromMany(type, out var elementType))
+            {
+                return elementType;
+            }
+            else
+            {
+                throw new ArgumentException(string.Format(Strings.ImportManyOnNonCollectionType, type.FullName), nameof(type));
+            }
+        }
+
+        internal static bool TryGetElementTypeFromMany(Type type, out Type elementType)
+        {
+            Requires.NotNull(type, nameof(type));
+            elementType = null;
+
```

Comment:
nit: move this to just above the `return false;` line. This will force anyone who returns `true` to remember to set this out parameter, whereas right now they could get away with forgetting and send null to the caller.